src/rust/engine/fs/src/lib.rs

```diff
@@ -764,36 +976,85 @@ pub trait VFS<E: Send + Sync + 'static>: Clone + Send + Sync + 'static {
     &self,
     path_glob: PathGlob,
     exclude: &Arc<Gitignore>,
-  ) -> BoxFuture<(Vec<PathStat>, Vec<PathGlob>), E> {
-    match path_glob {
+  ) -> BoxFuture<SingleExpansionResult, E> {
+    // If we were to match on &path_glob (which I just tried), it seems like it's impossible to move
+    // something while borrowing it, even if the move happens at the end of the borrowed scope. That
+    // seems like it doesn't need to happen? I think there is no way to avoid the clone here,
+    // however, because directory_listing() returns a future and therefore can use only static
+    // references.
+    match path_glob.clone() {
       PathGlob::Wildcard { canonical_dir, symbolic_path, wildcard } =>
-        // Filter directory listing to return PathStats, with no continuation.
+      // Filter directory listing to return PathStats, with no continuation.
         self.directory_listing(canonical_dir, symbolic_path, wildcard, exclude)
-          .map(|path_stats| (path_stats, vec![]))
-          .to_boxed(),
+        .map(move |path_stats| SingleExpansionResult {
+          path_glob,
+          path_stats,
+          globs: vec![],
+        })
+        .to_boxed(),
       PathGlob::DirWildcard { canonical_dir, symbolic_path, wildcard, remainder } =>
         // Filter directory listing and request additional PathGlobs for matched Dirs.
         self.directory_listing(canonical_dir, symbolic_path, wildcard, exclude)
-          .and_then(move |path_stats| {
-            path_stats.into_iter()
-              .filter_map(|ps| match ps {
-                PathStat::Dir { path, stat } =>
-                  Some(
-                    PathGlob::parse_globs(stat, path, &remainder)
-                      .map_err(|e| Self::mk_error(e.as_str()))
-                  ),
-                PathStat::File { .. } => None,
-              })
-              .collect::<Result<Vec<_>, E>>()
+        .and_then(move |path_stats| {
+          let new_globs = path_stats
+            .clone()
+            .into_iter()
+            .filter_map(|ps| match ps {
+              PathStat::Dir { path, stat } =>
+                Some(
+                  PathGlob::parse_glob(stat, path, &remainder)
+                    .map_err(|e| Self::mk_error(e.as_str()))
+                ),
+              PathStat::File { .. } => None,
+            })
+            .collect::<Result<Vec<_>, E>>();
+          future::result(new_globs)
+            .map(move |globs| SingleExpansionResult {
+              path_glob,
+              path_stats: vec![],
+              globs,
           })
-          .map(|path_globs| {
-            let flattened =
-              path_globs.into_iter()
-                .flat_map(|path_globs| path_globs.into_iter())
-                .collect();
-            (vec![], flattened)
+        })
+        .to_boxed(),
+      PathGlob::RecursiveWildcard { canonical_dir, symbolic_path, remainder } => {
```

Comment:
I'm not sure this new variant is worthwhile... it seems to add a lot of redundancy around `directory_listing`. Was it necessary in order to account for matches for some reason? If not, might be good to revert it.

In particular, recursion is already handled naturally by `loop_fn`, and so recursing here is redundant.

Reply:
I really wanted a single `String -> PathGlob`, not just because it really should be happening, but because then (with the `as_zsh_style_glob()` method I mentioned above) we can keep the 1:1 mapping between `input glob -> PathGlob`.